### ISSUE TITLE
Only getting one frame at 5000:/cam1 on a Raspberry pi 4

### ISSUE THREAD

**Issue by jalbstmeijer** [blakeblackshear/frigate#60]
Hi,

Got the frigate docker container running on my Raspberry pi 4 with a Coral TPU.
When visiting http://ip:5000/cam1 I only get 1 frame. The timestamp in the video stream stays the same.

I do not get any error, any idea?

Gr, Justin

**Comment by ivelin** [comment]
I have similar results on my RPI4 4GBRAM, but it does not bother me, because realistically a person or object that's not in front of the camera view for 2-3 seconds is not of much interest. 

With the latest rpi4 [optimizations](https://github.com/blakeblackshear/frigate/issues/59) that I posted, the CPU is down to 6% with one camera, which gives me a lot of bandwidth to add more cameras and more TF models. I will test next chaining of models, e.g. face detect after person detect.

**Comment by jalbstmeijer** [comment]
@ivelin 

With similar you mean you also only get 1 still frame from the stream at port 5000? but detection is still working for you? I was under the impression nothing got processed and therefore detection was also not working, but I'll have to look at that again.

Initially  I thought it might have something to do with the Coral libs.
From https://github.com/google-coral/edgetpu-platforms you get the impression that Raspberry pi 4 is not yet supported by the default edgetpu_api_2.11.1.tar.gz from https://coral.withgoogle.com/resources/

But you are also just using the default edgetpu_api_2.11.1.tar.gz from https://coral.withgoogle.com/resources/ ?

btw, I had to omit the i965-va-driver fro the Dockerfile to get the image build on my Raspberry pi 4.

**Comment by ivelin** [comment]
> @ivelin

Let me clarify. You make a few good points.

> 
> With similar you mean you also only get 1 still frame from the stream at port 5000? but detection is still working for you? 

Yes. Detection is still working for me. Do you see the region boxes drown in white border color and red border boxes for detections? Also do you see /best_person.jpg when someone is in front of the camera?

> I was under the impression nothing got processed and therefore detection was also not working, but I'll have to look at that again.

If you don't get the detection, then we need to get that sorted out first.

> 
> Initially I thought it might have something to do with the Coral libs.

Yes, you have to make build config changes and rebuild Coral libs for RPI4. The default binary only works on RPI3. I can help you with that if you are stuck here.

> From https://github.com/google-coral/edgetpu-platforms you get the impression that Raspberry pi 4 is not yet supported by the default edgetpu_api_2.11.1.tar.gz from https://coral.withgoogle.com/resources/
> 
> But you are also just using the default edgetpu_api_2.11.1.tar.gz from https://coral.withgoogle.com/resources/ ?
> 
> btw, I had to omit the i965-va-driver from the Dockerfile to get the image build working on my Raspberry pi 4.

Yes, same here.

If you like, you can try the unofficial frigate docker image for rpi4 that I shared. Its based on the latest source and should include the Coral binaries.
https://hub.docker.com/r/ivelin/frigate

`docker pull ivelin/frigate:unofficial_pi4_aug_20_2019
`

**Comment by jalbstmeijer** [comment]
Hi, tried ivelin/frigate:unofficial_pi4_aug_20_2019.
Same result. No stream at port 5000 and nothing detected.
Tried with the example regions from https://github.com/blakeblackshear/frigate/blob/master/config/config.yml which seem to result quickly in this message 'queue full. moving on' but also with tried

```
    regions:
      - size: -1
        x_offset: 0
        y_offset: 0
        min_person_area: 5000
        threshold: 0.5

```

**Comment by jalbstmeijer** [comment]
When I make sure a person is before the camera when starting the container, that person gets detected. After that not anymore.

Cpu usage is very high, specially by Python

I'm using a Foscam Fi9831W camera and lowered resolution now to 640x480 just to be sure.

**Comment by jalbstmeijer** [comment]
Interesting, it seems to work when I connect the Coral to the upper left usb 2 port (black).
Cpu usage lowers and even the stream on port 5000 now works.

**Comment by ivelin** [comment]
That's interesting. Are you able to connect to USB3 port. It can speed
things up 5x.

**Comment by jalbstmeijer** [comment]
@ivelin , are you also running Raspbian Buster with kernel 4.19.66-v7l+ and VL805 FW version 00013701?

It seems it works accidentally.
Currently the most reproducible situation is "queue full. moving on" when starting the container.
 
```
On connect called
INFO: Initialized TensorFlow Lite runtime.
Creating a new ffmpeg process...
ffmpeg -hide_banner -loglevel panic -avoid_negative_ts make_zero -fflags nobuffer -flags low_delay -strict experimental -fflags +genpts -rtsp_transport tcp -stimeout 5000000 -use_wallclock_as_timestamps 1 -i rtsp://admin:XXXXXX@192.168.xx.xx:88/videoSub -f rawvideo -pix_fmt rgb24 pipe:
Creating a new capture thread...
Starting a new capture thread...
Capture process for cam1: 31
 * Serving Flask app "detect_objects" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://0.0.0.0:5000/ (Press CTRL+C to quit)
queue full. moving on
queue full. moving on
queue full. moving on

```

**Comment by ivelin** [comment]
Interesting. I wonder what's different on my RPI4. Looks the same.

Let me push a PR that I'm in the middle of and will take a closer look at
your settings.

**Comment by ivelin** [comment]
One more thing. Because of the use of DetectWithImage, it is no longer necessary for the regions to be square. The function resizes to fit the model input tensor while keeping the ratio of the fed image area. 
If its OK, I would suggest allowing regions to be optional. By default the whole image can be passed to the model. Further zooming in can take place by model chaining from more generic to more specialized models.

**Comment by ivelin** [comment]
@jalbstmeijer 
Looking at the error messages closer, my guess is that the thread processing captured images via Tensorflow (PreppedQueueProcessor) dies quietly and doesn't pull images from the queue.

If you have the time to dive in the python source and debug, I think we can figure it out. 

Can you add a few try/catch/debug prints in the run method of PreppedQueueProcessor. Especially around the line that calls Coral via DetectWith* .

**Comment by ivelin** [comment]
> 4.19.66

I was running on 4.19.58. Upgrading now to see if I can reproduce your error.

**Comment by jalbstmeijer** [comment]
> Can you add a few try/catch/debug prints in the run method of PreppedQueueProcessor. Especially around the line that calls Coral via DetectWith* .

Sounds as the right approach, but my Python debugging experience is non existing.
So I will give it a try,  but if you would have debug code to add it would much appreciated.

**Comment by ivelin** [comment]
Try this around the line DetectWith*:

```
            try:
                print("Calling TF Coral to detect objects")
                objects = self.engine.DetectWith...
                print("TF Coral detected {} objects".format(len(objects)))
            except Exception as e:
                print("Error while calling TF Coral: \n" + str(e))
```

I was able to reproduce `queue full. moving on` by unplugging the Coral USB cord at runtime. The debug code above resulted in the following exception:

```
HandleQueuedBulkIn transfer in failed. Not found: USB transfer error 5 [LibUsbDataInCallback]
```

However when I run frigate with Coral unplugged, it fails earlier with
```
  File "/opt/frigate/frigate/object_detection.py", line 42, in __init__
    self.engine = DetectionEngine(tf_model)
  File "/usr/local/lib/python3.6/dist-packages/edgetpu/detection/engine.py", line 72, in __init__
    super().__init__(model_path)
  File "/usr/local/lib/python3.6/dist-packages/edgetpu/basic/basic_engine.py", line 40, in __init__
    self._engine = BasicEnginePythonWrapper.CreateFromFile(model_path)
RuntimeError: No Edge TPU device detected!
```

So in your case, it seems like you pass the Coral engine setup stage 

`        self.engine = DetectionEngine(tf_model)
`
but something bombs during detection.

Let us know what exception you get.

**Comment by jalbstmeijer** [comment]
@ivelin , no exception at this level;

> Calling TF Coral to detect objects
> TF Coral detected 0 objects
> Calling TF Coral to detect objects
> TF Coral detected 0 objects
> Calling TF Coral to detect objects
> queue full. moving on
> queue full. moving on
> TF Coral detected 0 objects
> Calling TF Coral to detect objects
> queue full. moving on
> queue full. moving on
> queue full. moving on
>

**Comment by jalbstmeijer** [comment]
Just double checked, inside the frigate container, if Coral was just not responding, but it seems to be able to run the parrot demo:

> root@3ab3ee06f38c:/usr/local/lib/python3.6/dist-packages/edgetpu/demo# while true; do time python3 classify_image.py --model ~/Downloads/mobilenet_v2_1.0_224_inat_bird_quant_edgetpu.tflite --label ~/Downloads/inat_bird_labels.txt --image ~/Downloads/parrot.jpg; done
> INFO: Initialized TensorFlow Lite runtime.
> ---------------------------
> Ara macao (Scarlet Macaw)
> Score :  0.76171875
> 
> real	0m3.432s
> user	0m0.508s
> sys	0m0.204s
> INFO: Initialized TensorFlow Lite runtime.
> ---------------------------
> Ara macao (Scarlet Macaw)
> Score :  0.76171875
> 
> real	0m3.398s
> user	0m0.489s
> sys	0m0.190s
> INFO: Initialized TensorFlow Lite runtime.
> ---------------------------
> Ara macao (Scarlet Macaw)
> Score :  0.76171875
> 
> real	0m3.414s
> user	0m0.517s
> sys	0m0.180s
> INFO: Initialized TensorFlow Lite runtime.
> ---------------------------
> Ara macao (Scarlet Macaw)
> Score :  0.76171875
> 
> real	0m3.390s
> user	0m0.493s
> sys	0m0.182s
> INFO: Initialized TensorFlow Lite runtime.
> ---------------------------
> Ara macao (Scarlet Macaw)
> Score :  0.76171875
> 
> real	0m3.397s
> user	0m0.532s
> sys	0m0.144s
>

**Comment by ivelin** [comment]
OK, back to the drawing board :)

If I understood your earlier comments correctly, you only ever see one image in the browser. The clock in the image never advances? If that is the case, then we can look next at the image feed thread in `detect_objects.py`.

Can you add a similar try/exception block below and prints as we did earlier.

```

    def imagestream(camera_name):
        while True:
            # max out at 5 FPS
            time.sleep(0.2)
            try:
              frame = cameras[camera_name].get_current_frame_with_objects()
              # encode the image into a jpg
              ret, jpg = cv2.imencode('.jpg', frame)
              yield (b'--frame\r\n'
                  b'Content-Type: image/jpeg\r\n\r\n' + jpg.tobytes() + b'\r\n\r\n')
           except Exception as e:
                print("Error while preparing the next camera image for http output: \n" + str(e))

```


If this doesn't help much I am open to setup a screenshare session. I have flexibility in the next few days.

**Comment by jalbstmeijer** [comment]
@ivelin nothing extra gets logged with the additional lines.
I'm happy to supply access to my PI4.
Let's communicate directly to exchange access details.

**Comment by jalbstmeijer** [comment]
> If I understood your earlier comments correctly, you only ever see one image in the browser. The clock in the image never advances? If that is the case, then we can look next at the image feed thread in detect_objects.py.
Correct!

**Comment by jalbstmeijer** [comment]
ivelin was so friendly to debug my issue.
eventually he got it to work by tuning ffmpeg settings.

Before I was just using the frigate code base that was included in the Docker image,  while debugging we mounted a separate code folder to make editing easier. 

After the debug session I realized this mounted code folder was also from a couple of weeks ago.

To make a long story story. A fresh git clone without any modifications now works fine for me.
I can reproduce my issues when switching back to 0.2.0, so it seems it is fixed in 0.2.1 for me.
